rfcs/search_relevancy_sandbox/20230406-implementation_plan_update_staging_database.md

```diff
@@ -0,0 +1,166 @@
+   [`restore_db_instance_from_db_snapshot`](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/rds/client/restore_db_instance_from_db_snapshot.html)
+   function (an operator does not exist for this operation). This database will
+   be named in a way that does not conflict with the existing dev database name,
+   e.g. `dev-next-openverse-db`. Careful consideration will need to be made when
+   drafting this step to ensure that the generated database matches the settings
+   for the existing database exactly. Specifically, the following should be
+   matched:
+   - Availability should be a single DB instance (not multi-AZ)
+   - It should be attached to our default VPC
+   - It should not be publicly accessible
+   - It will need access to the following VPC security groups: `default`,
+     `openverse-vpc-db-access`, `staging-dev-rds-sg`
+   - It should match the dev database's instance size (currently `m5.xlarge`)
+   - It should have 3000 GB of allocated gp2 storage
+   - It should use password authentication (this may not need to be changed from
+     the default, as the snapshot should contain the same connection
+     information)
+4. Wait for the new database to be ready using the
+   [`RdsDbSensor`](https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/_api/airflow/providers/amazon/aws/sensors/rds/index.html#airflow.providers.amazon.aws.sensors.rds.RdsDbSensor)
+   ([example](https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/operators/rds.html#howto-sensor-rdsdbsensor)).
+5. Rename the old staging database to `dev-old-openverse-db` using boto3's
+   [`modify_db_instance`](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/rds/client/modify_db_instance.html)
+   function (an operator does not exist for this operation).
+6. Wait for the old database rename to be complete with the `RdsDbSensor` (we
+   may need retries on this step, since the database may not be initially
+   available/named when the sensor first starts). _**Note**: this will cause a
+   temporary outage of the staging API, see
+   [the alternatives section](#alternatives) for why this is the case._
+7. Rename the new database to `dev-openverse-db` using `modify_db_instance`.
+8. Wait for the new database rename to be complete with the `RdsDbSensor` (we
+   may need retries on this step, since the database may not be initially
+   available/named when the sensor first starts).
+9. If the previous steps fail, rename `dev-old-openverse-db` back to
+   `dev-openverse-db`. Otherwise, `dev-old-openverse-db` can be deleted using
+   the
+   [`RdsDeleteDbInstanceOperator`](https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/_api/airflow/providers/amazon/aws/operators/rds/index.html#airflow.providers.amazon.aws.operators.rds.RdsDeleteDbInstanceOperator)
+   (the `wait_for_completion` should be left as `True` here so we don't need a
+   follow-up sensor).
```

Comment:
I think my wording may be ambiguous here - the `wait_for_completion` step means that we avoid the need for a sensor _on the delete operation itself_. The `wait_for_completion` parameter is a feature on the Airflow operators themselves, unfortunately anything we do with the RDS client directly will require a follow-up sensor since I don't _think_ there's a similar mechanism available with the `boto3` API alone. I'll reword this!

Reply:
OH got it, that did totally confuse me -- I was thinking of `wait_for_completion` as the task id for a sensor. Got it!